core/modules/css/KvInterfaceImplMySql.cc

```diff
@@ -0,0 +1,319 @@
+        }
+    }
+
+    /// Explicitly abort transaction, throws SqlError for errors.
+    void abort() {
+        _trans.abort(_errObj);
+        if (_errObj.isSet()) {
+            LOGF(_logger, LOG_LVL_ERROR, "Failed to abort transaction: mysql error: (%1%) %2%" %
+                 _errObj.errNo() % _errObj.errMsg());
+            throw CssError(_errObj);
+        }
+    }
+
+private:
+    sql::SqlErrorObject _errObj; // this must be declared before _trans
+    sql::SqlTransaction _trans;
+};
+
+
+KvInterfaceImplMySql::KvInterfaceImplMySql(mysql::MySqlConfig const& mysqlConf)
+: _conn(mysqlConf) {
+
+}
+
+
+void
+KvInterfaceImplMySql::create(std::string const& key, std::string const& value) {
+    if (key.size() > kMaxKeyLength) {
+        LOGF(_logger, LOG_LVL_DEBUG, "create - rejecting; key too long: %1%" % key);
+        throw CssError("key too long");
+    }
+
+    KvTransaction transaction(_conn);
+
+    boost::filesystem::path keyPathObj(key);
+    boost::filesystem::path parentKeyPathObj = keyPathObj.parent_path();
+
+    std::string parentKey = boost::filesystem::path(key).parent_path().string();
+
+    boost::format fmQuery("INSERT INTO cssData (kvKey, kvVal, parentKey) VALUES ('%1%', '%2%', '%3%');");
+    fmQuery % key;
+    fmQuery % value;
+    fmQuery % parentKey;
+    std::string query = fmQuery.str();
+
+    // run query
+    sql::SqlErrorObject errObj;
+    LOGF(_logger, LOG_LVL_DEBUG, "create - executing query: %1%" % query);
+    if (not _conn.runQuery(query, errObj)) {
+        switch (errObj.errNo()) {
+        default:
+            throw CssError(errObj);
+            break;
+
+        case ER_DUP_ENTRY:
+            LOGF(_logger, LOG_LVL_ERROR, "create - SQL INSERT INTO failed: %1%" % query);
+            throw KeyExistsError(errObj);
+        }
+    }
+
+    transaction.commit();
+}
+
+
+void
+KvInterfaceImplMySql::set(std::string const& key, std::string const& value) {
+    if (key.size() > kMaxKeyLength) {
+        LOGF(_logger, LOG_LVL_DEBUG, "set - rejecting; key too long: %1%" % key);
+        throw CssError("key too long");
+    }
+
+    KvTransaction transaction(_conn);
+
+    boost::format fmQuery("INSERT INTO cssData (kvKey, kvVal) VALUES ('%1%', '%2%') ON DUPLICATE KEY UPDATE kvVal='%2%';");
+    fmQuery % key;
+    fmQuery % value;
+    std::string query = fmQuery.str();
+
+    // run query
+    sql::SqlErrorObject errObj;
+    LOGF(_logger, LOG_LVL_DEBUG, "set - executing query: %1%" % query);
+    if (not _conn.runQuery(query, errObj)) {
+        LOGF(_logger, LOG_LVL_ERROR, "%1% failed with err:%2%" % query % errObj.errMsg());
+        throw CssError(errObj);
+    } else if (errObj.isSet()) {
+        throw CssError(errObj);
+    }
+
+    transaction.commit();
+}
+
+
+bool
+KvInterfaceImplMySql::exists(std::string const& key) {
+    KvTransaction transaction(_conn);
+
+    std::string val;
+    sql::SqlErrorObject errObj;
+    return _getValFromServer(key, val, errObj);
```

Comment:
`_getValFromServer` can return false when either the key is missing or something is wrong on mysql side. You probably want to look at `errObj` and throw an exception if it has an error condition set.